tokio-util/src/io/sink_writer.rs

```diff
@@ -0,0 +1,183 @@
+use bytes::Bytes;
+use futures_sink::Sink;
+
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::AsyncWrite;
+
+pin_project! {
+    /// Convert a [`Sink`] of byte chunks into an [`AsyncWrite`].
+    /// Bytes are sent into the sink and the sink is flushed once all bytes are sent. If an error occurs during the sending progress,
+    /// the number of sent but unflushed bytes are saved in case the flushing operation stays unsuccessful.
+    /// For the inverse operation of defining an [`AsyncWrite`] from a [`Sink`] you need to define a [`codec`].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use futures_util::SinkExt;
+    /// use std::io::{Error, ErrorKind};
+    /// use tokio::io::AsyncWriteExt;
+    /// use tokio_util::io::{SinkWriter, CopyToBytes};
+    /// use tokio_util::sync::PollSender;
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() -> Result<(), Error> {
+    ///  // Construct a channel pair to send data across and wrap a pollable sink.
+    ///  // Note that the sink must mimic a writable object, e.g. have `std::io::Error`
+    ///  // as its error type.
+    /// let (tx, mut rx) = tokio::sync::mpsc::channel::<Vec<u8>>(1);
+    ///  let mut writer = SinkWriter::new(CopyToBytes::new(
+    ///    PollSender::new(tx).sink_map_err(|_| Error::from(ErrorKind::BrokenPipe)),
+    /// ));
+    ///  // Write data to our interface...
+    ///  let data: [u8; 4] = [1, 2, 3, 4];
+    ///  let _ = writer.write(&data).await?;
+    ///  // ... and receive it.
+    ///  assert_eq!(data.to_vec(), rx.recv().await.unwrap());
+    ///
+    /// #  Ok(())
+    /// # }
+    /// ```
+    ///
+    ///
+    /// [`AsyncWrite`]: tokio::io::AsyncWrite
+    /// [`Sink`]: futures_sink::Sink
+    /// [`codec`]: tokio_util::codec
+    #[derive(Debug)]
+    pub struct SinkWriter<S>
+    {
+        #[pin]
+        inner: S,
+    }
+}
+
+impl<S> SinkWriter<S> {
+    /// Creates a new [`SinkWriter<S>`].
+    pub fn new(sink: S) -> Self {
+        Self { inner: sink }
+    }
+
+    /// Gets a reference to the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn get_ref(&self) -> &S {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.inner
+    }
+
+    /// Consumes this `SinkWriter`, returning the underlying sink.
+    ///
+    /// It is inadvisable to directly write to the underlying sink.
+    pub fn into_inner(self) -> S {
+        self.inner
+    }
+}
+pin_project! {
+    /// A helper structure, converting a sink of byte slices into a
+    /// sink of owned [`Bytes`].
```

Comment:
The documentation here isn't sufficient. Can you add some more details? You should explain that this is a _wrapper_ and that it can be used to convert a `Sink<Bytes>` into a `Sink<&'a [u8]>`.

Reply:
I updated it a bit. What do you think?